### PR TITLE
Include superclasses when searching for model properties

### PR DESCRIPTION
**Pull request by samalone**
Improves `AnyModel.anyID` and `AnyModel.properties` so they can find model properties in superclasses. This is needed to allow class inheritance in models like:

```swift
class Foo: Model {
    class var schema: String { "foo" }

    @ID(key: "id")
    var id: Int?
    
    required init() { }
}

class Bar: Foo {
    class override var schema: String { "bar" }
    
    @Field(key: "value")
    var value: String
}
```

### CODE REVIEW

**Comment by samalone** [comment]
(I should say up front that the last time I claimed that Swift couldn't do something, I turned out to be wrong, so take this with a grain of salt...)

When I tried to use protocol-oriented programming with Fluent, I ran into a roadblock. Swift protocols cannot declare property wrappers, and Fluent requires access to property wrappers to modify relationships. This means that protocols cannot modify relationships between Fluent objects, and so are too limited to be of much use.

In my case, I was trying to set up a versioned database where each modification is stored as a new data record, and a master record points to the current version of the data record. Since the database will need several type of data, this pattern needs to be repeated over and over again for different record types. (This is a simplification of the actual design, which actually has 3 tables for each type of data.)

The whole system depends on modifying the relationship between records in different tables, and I couldn't find any way to do this with protocols. The only way I have been successful was by creating base classes for each of the 3 table types, and then using generic classes to specialize them to store particular data. But that solution requires class inheritance, which is the point of this pull request.

If Fluent were able to update relationships simply by calling the standard property setter, protocols would work fine. But until then, class inheritance is the only way I have found to reuse code that creates or modifies relationships.

**Comment by tanner0101** [comment]
Yeah Swift's lack of support for property wrappers in protocols is unfortunate. I hope this is added in the future. 

Until then, this is a useful workaround:

```swift
protocol Foo {
    // @Field
    var bar: Int { get set }
}

extension Foo {
    var _$bar: Field<Int> {
            guard let bar = Mirror(reflecting: self).descendant("_bar") as? Field<Int> else {
                fatalError("bar property must be declared using @Field")
            }
            return bar
    }
}
```

**Comment by samalone** [comment]
LOL. Using reflection to avoid inheritance isn't a solution I would have come up with. 😀

That's a very helpful technique. Thanks for passing it on. I still feel like class inheritance is a cleaner solution than runtime symbol lookups, but that's a matter of opinion and it **_is_** nice to be able to use protocols with Fluent. I may refactor my own code to use this technique.

I still think it's worth supporting class inheritance in a framework that will be used by diverse programmers. Not everyone using Fluent will have the same level of comfort creating protocol-oriented designs, so I think it's worth supporting class inheritance as well if there are no technical obstacles.

I'll also point out that it's possible to support inheritance and still require final classes if that's needed to make Fluent work. In my own case, only the final classes are stored in the database. The base classes are only for code re-use.

**Comment by tanner0101** [comment]
I see your point. I think getting inheritance to work throughout Fluent would be significantly more work than this PR has though. All of the code written so far has assumed no inheritance. For example, would this extension still work? https://github.com/vapor/fluent-kit/blob/master/Sources/FluentKit/Model/Model.swift#L229-L234. We'd ideally be testing every feature Fluent has with inheritance based models and I'm not sure that's worth it when protocols can do all the same things.

**Comment by samalone** [comment]
Umm... That's precisely one of the two pieces of code modified by my pull request.

https://github.com/vapor/fluent-kit/pull/148/commits/646f02edb5ee7906e87c498fe23a7207270a35a9

**Comment by samalone** [comment]
I should also mention that when I originally developed my pull request, I searched the fluent-kit source code for all calls to `Mirror(reflecting:)`. There were only two occurrences, which are the places I changed. As best as I can tell, all code that needs property access goes through either `AnyModel.properties` or `AnyModel.anyID`.

**Comment by tanner0101** [comment]
D'oh, sorry I misread that. Those do indeed seem to be the only places using `Mirror` so maybe this could work. I still feel like there may be some underwater rocks but we could always walk this back if we end up hitting those.

File: Sources/FluentKit/Model/Model.swift [hunk 1]
```diff
@@ -227,10 +227,16 @@ extension AnyModel {
...
     var anyID: AnyID {
-        guard let id = Mirror(reflecting: self).descendant("_id") as? AnyID else {
-            fatalError("id property must be declared using @ID")
+        var mirror: Mirror = Mirror(reflecting: self)
+        while true {
+            if let id = mirror.descendant("_id") as? AnyID { return id }
+            if let m = mirror.superclassMirror {
+                mirror = m
+            }
+            else {
```
Comment (tanner0101) [inline Sources/FluentKit/Model/Model.swift]: Can you put the elses on the same line as the closing `}`?

File: Sources/FluentKit/Properties/Property.swift [hunk 2]
```diff
@@ -78,18 +78,27 @@ extension AnyModel {
...
+                    guard let label = child.label else {
+                        return nil
+                    }
+                    guard let property = child.value as? AnyProperty else {
+                        return nil
+                    }
+                    // remove underscore
+                    return (String(label.dropFirst()), property)
+            })
+            if let m = mirror.superclassMirror {
```
Comment (tanner0101) [inline Sources/FluentKit/Properties/Property.swift]: Could this be written as:
```swift
var currentMirror: Mirror? = Mirror(reflecting: self)
while let mirror = currentMirror {
    // add props
    currentMirror = mirror.superclassMirror
}
```

Comment (samalone) [reply Sources/FluentKit/Properties/Property.swift]: Do you like this any better?

```swift
var mirror: Mirror = Mirror(reflecting: self)
while true {
  // add props
  guard let nextMirror = mirror.superclassMirror else { return props }
  mirror = nextMirror
}
```

File: Sources/FluentKit/Model/Model.swift (hunk 1 above)
Comment (samalone) [reply Sources/FluentKit/Model/Model.swift]: No problem. The next commit will do this in both places.

File: Sources/FluentKit/Properties/Property.swift (hunk 2 above)
Comment (tanner0101) [reply Sources/FluentKit/Properties/Property.swift]: That's a little better. `while true` is a little confusing to me though. When I see that I think this loop should be going forever. Putting the stop condition in the `while` makes it clearer what the loop is doing.

File: Sources/FluentKit/Properties/Property.swift [hunk 3]
```diff
@@ -78,18 +78,23 @@ extension AnyModel {
...
-                }
-                guard let property = child.value as? AnyProperty else {
-                    return nil
-                }
-                // remove underscore
-                return (String(label.dropFirst()), property)
-            }
+        var props: [(String, AnyProperty)] = []
+        var currentMirror: Mirror? = Mirror(reflecting: self)
+        while let mirror = currentMirror {
```
Comment (tanner0101) [inline Sources/FluentKit/Properties/Property.swift]: +1 this is a lot clearer.

File: Sources/FluentKit/Model/Model.swift [hunk 4]
```diff
@@ -227,10 +227,15 @@ extension AnyModel {
     }
 
     var anyID: AnyID {
-        guard let id = Mirror(reflecting: self).descendant("_id") as? AnyID else {
-            fatalError("id property must be declared using @ID")
+        var mirror: Mirror = Mirror(reflecting: self)
+        while true {
```
Comment (tanner0101) [inline Sources/FluentKit/Model/Model.swift]: Same thing here.

**Comment by tanner0101** [comment]
Closing for inactivity, feel free to re-open with updates, thanks.

**Pull request closed without merging**